### ISSUE TITLE
How to check that communication is coming from president?

### ISSUE THREAD

**Issue by mpardalos** [SOMAS2020/SOMAS2020#164]
Since anyone can send communications to anyone, you need a way to check that "official" communications are coming from the correct agent (i.e. the current president/judge/speaker) is there any way to get the ID of the current e.g. president, judge and speaker?

I'm trying to write this code:

```go
func (c *client) ReceiveCommunication(
	sender shared.ClientID,
	data map[shared.CommunicationFieldName]shared.CommunicationContent,
) {
	for field, content := range data {
		switch field {
		case shared.TaxAmount:
			if __isPresident__(sender) {
				c.taxAmount = shared.Resources(content.IntegerData)
			} else {
				// Someone is impersonating the president
			}
		}
	}
}
```

What would go into `__isPresident__(sender)`?

I am simply trying to pay my taxes like a law abiding citizen and not get scammed.

**Comment by mpardalos** [comment]
CC: @Neelesh99 @preetl

**Comment by Neelesh99** [comment]
This is something that we will add to the global variable cache for post MVP, currently (and quite crudely) it's global in the iigointernal package, we will remove that in post-MVP. If you think there's a better/more secure way of storing this info pls let us know.

**Comment by lhl2617** [comment]
# TTP method

the server can be modelled as a TTP (Trusted Third Party) to verify the identity of a message sender.

each turn the server generates a random number `p` for the president.

the server (defined in the `ServerReadHandle`) has a cryptographic hash fn `f(p, t)` where `p` is as above, `t` is the target client ID

the president, when sending a message to `t`, calls the `f(p, t)` function to get `f'`. (optionally) the server stores the mapping `t -> f(p, t)`.

the recipient `t` can call a verification function on the `ServerReadHandle` to verify that `f'` is indeed generated by a president.

this is barely computationally secure (based on the cryptographic hash fn `f`). the same malicious attacks on the `ServerReadHandle` server ptr can compromise the game.

# Via server method

easier, just send comms via the server and the server can tag each message with a verified role.

**Comment by preetl** [comment]
@lhl2617 I see the merit in your proposal but I would say that it is safe to assume that no team is going to be coding maliciously and making agents to take advantage of the infrastructure we have put in place. I think it is not necessary to go to such lengths to have secure communication. In my opinion, it is sufficient to have the president, speaker and judge role ids accessible but read only for the agents i.e. by putting them in clientgamestate so they can read them through the serverreadhandle. Islands can corroborate that messages are coming from the president by checking the sender id against this source.

If this is agreeable to you @mpardalos @lhl2617 @Neelesh99, I'll put the IIGO role ids into clientgamestate.

**Comment by mpardalos** [comment]
I'm with @preetl here. I think the best way to do this is to just have the officials' IDs in the `ClientGameState`. The sender is set by the server anyway so I don't really see how a client could maliciously/erroneously mess with that.

**Comment by lhl2617** [comment]
Ah fairs! Yep, it's what I meant with the "via server method"

**Comment by preetl** [comment]
Solved in PR #226
